### PR TITLE
[AKS] `az aks create/update`: Grafana support, update region map and add cluster scope

### PR DESCRIPTION
**Pull request by bragi92**
**Related command**

```az aks create -n kaveeshcli22 -g kaveeshcli --location westeurope --enable-azure-monitor-metrics --azure-monitor-workspace-resource-id "{full_id}" --grafana-resource-id "{full_id}"```

```az aks update -n kaveeshcli22 -g kaveeshcli --enable-azure-monitor-metrics --enable-windows-recording-rules```

```az aks create -n kaveeshcli22 -g kaveeshcli --disable-azure-monitor-metrics ```

**Description**

- Enable grafana support in Fairfax
- Update region map for default region creation with new Azure Monitor Workspace regions
- Add cluster scope to recording rule groups created during Managed Prometheus onboardinng
- Update DC* object names to match the azure portal
- Add subscription scope to get_supported_rp_locations for Microsoft.Monitor RP to respect feature flags the customer might've set
- Limit recording rule group name length to 260 characters (creation/deletion) 

DC* refers to [DCRA](https://learn.microsoft.com/en-us/rest/api/monitor/data-collection-rule-associations), [DCR ](https://learn.microsoft.com/en-us/rest/api/monitor/data-collection-rules) and [DCE](https://learn.microsoft.com/en-us/rest/api/monitor/data-collection-endpoints) i.e. Data Collection Rule Association, Data Collection Rules and Data Collection Endpoints that are being created as part of the managed prometheus addon onboarding.

**Testing Guide**

**Scenario (grafana support in Fairfax)**:
- Set context to AzureUSGovernment `az cloud set --name AzureUSGovernment` 
- Use the command with the grafana-resource-id parameter similar to `az aks create -n kaveeshcli22 -g kaveeshcli --location westeurope --enable-azure-monitor-metrics --azure-monitor-workspace-resource-id "{full_id}" --grafana-resource-id "{full_id}"` and it should result in a succees

**Scenario (region map update)**:
- Set context to AzureCloud `az cloud set --name AzureCloud` 
- Create a new cluster without specifying an Azure Monitor Workspace -> `az aks create -n kaveeshcli22 -g kaveeshcli --location italynorth--enable-azure-monitor-metrics`
- Use a cluster in a new region that is being updated in this PR like -> italynorth, israelcentral or germanywestcentral. Make sure that the onboarding succeeds and the output contains a line similar to the following -> Using Azure Monitor Workspace (stores prometheus metrics) : DefaultAzureMonitorWorkspace-italynorth

**Scenario (cluster scope update)**:
- Set context to AzureCloud `az cloud set --name AzureCloud` 
- Use the command `az aks create -n kaveeshcli22 -g kaveeshcli --location westeurope --enable-azure-monitor-metrics --azure-monitor-workspace-resource-id "{full_id}" --grafana-resource-id "{full_id}"` and it should result in a succees

**History Notes**

[AKS] `az aks create/update`: Enable Grafana support in US Government clouds
[AKS] `az aks create/update`: Update region map for default region creation with new Azure Monitor Workspace regions
[AKS] `az aks create/update`: Add cluster scope to recording rule groups created during Managed Prometheus onboarding

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Thank you for your contribution! We will review the pull request and get back to you soon.

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/dc/dcr_api.py [hunk 1]
```diff
@@ -3,27 +3,17 @@
...
 # --------------------------------------------------------------------------------------------
 import json
-from azure.cli.command_modules.acs.azuremonitormetrics.constants import MapToClosestMACRegion
-from azure.cli.command_modules.acs.azuremonitormetrics.dc.defaults import get_default_region, sanitize_name
+from azure.cli.command_modules.acs.azuremonitormetrics.dc.defaults import get_default_dcr_name
 from azure.cli.command_modules.acs.azuremonitormetrics.constants import (
-    DC_TYPE,
     DC_API
 )
 from knack.util import CLIError
```
Comment (bragi92) [inline src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/dc/dcr_api.py:10]: Add subscription scope to get_supported_rp_locations

Comment (bragi92) [reply src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/dc/dcr_api.py:10]: done

**Comment by yanzhudd** [comment]
Hi @bragi92, 
since there are more than one changes in this PR, please list them respectively in the [History Notes](https://github.com/Azure/azure-cli/pull/27488#issue-1916386198) (format please refer to the examples).

**Comment by yanzhudd** [comment]
Please add test for these changes.

**Comment by bragi92** [comment]
@yanzhudd  Please take a look at this PR when you get a chance. I have added unit tests and updated the description as requested.

**Comment by yonzhan** [comment]
Please fix CI issues.

**Comment by bragi92** [comment]
@yonzhan It seems like all builds are failing right now in the CLI pipeline with a build issue in 'Test Extensions Loading Python311) and this is not related to the changes in my PR at all.

**Comment by zhoxing-ms** [comment]
@bragi92 Could you please add some scenario test for this PR?

**Comment by zhoxing-ms** [comment]
@FumingZhang Could you please help review this AKS related PR as well?

**Comment by zhoxing-ms** [comment]
@bragi92 Please note that we are launching the release for this sprint this week. Please resolve all comments and conflicts by Tomorrow, otherwise the release of this PR will have to be postponed to the next sprint **(on 12-05)**

**Comment by bragi92** [comment]
@zhoxing-ms I believe I have addressed the PR comments already. Please let me know if anything else is remaining.

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/create.py [hunk 2]
```diff
@@ -50,23 +51,30 @@ def put_rules(cmd, default_rule_group_id, default_rule_group_name, mac_region, a
 
 # pylint: disable=line-too-long
 def create_rules(cmd, cluster_subscription, cluster_resource_group_name, cluster_name, azure_monitor_workspace_resource_id, mac_region, raw_parameters):
+    # limit rule group name to 260 characters
     # with urllib.request.urlopen("https://defaultrulessc.blob.core.windows.net/defaultrules/ManagedPrometheusDefaultRecordingRules.json") as url:
     #     default_rules_template = json.loads(url.read().decode())
     default_rules_template = get_recording_rules_template(cmd, azure_monitor_workspace_resource_id)
-    default_rule_group_name = "NodeRecordingRulesRuleGroup-{0}".format(cluster_name)
+    default_rule_group_name = "NodeRecordingRulesRuleGroup-{0}".format(cluster_name[:232])
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/create.py:58]: use a function to truncate the concatenated name to a maximum of 260 characters would be more consistent in my opinion

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/create.py [hunk 3]
```diff
@@ -50,23 +51,30 @@ def put_rules(cmd, default_rule_group_id, default_rule_group_name, mac_region, a
...
     url = "{0}{1}?api-version={2}".format(
         cmd.cli_ctx.cloud.endpoints.resource_manager,
         default_rule_group_id,
         RULES_API
     )
-    put_rules(cmd, default_rule_group_id, default_rule_group_name, mac_region, azure_monitor_workspace_resource_id, cluster_name, default_rules_template, url, True, 0)
+    put_rules(cmd, default_rule_group_id, default_rule_group_name, mac_region, cluster_resource_id, azure_monitor_workspace_resource_id, cluster_name, default_rules_template, url, True, 0)
 
-    default_rule_group_name = "KubernetesRecordingRulesRuleGroup-{0}".format(cluster_name)
+    default_rule_group_name = "KubernetesRecordingRulesRuleGroup-{0}".format(cluster_name[:226])
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/create.py:77]: use a helper function for truncation?

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/create.py [hunk 4]
```diff
@@ -77,14 +85,14 @@ def create_rules(cmd, cluster_subscription, cluster_resource_group_name, cluster
...
-    put_rules(cmd, default_rule_group_id, default_rule_group_name, mac_region, azure_monitor_workspace_resource_id, cluster_name, default_rules_template, url, True, 1)
+    put_rules(cmd, default_rule_group_id, default_rule_group_name, mac_region, cluster_resource_id, azure_monitor_workspace_resource_id, cluster_name, default_rules_template, url, True, 1)
 
     enable_windows_recording_rules = raw_parameters.get("enable_windows_recording_rules")
 
     if enable_windows_recording_rules is not True:
         enable_windows_recording_rules = False
 
-    default_rule_group_name = "NodeRecordingRulesRuleGroup-Win-{0}".format(cluster_name)
+    default_rule_group_name = "NodeRecordingRulesRuleGroup-Win-{0}".format(cluster_name[:228])
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/create.py:95]: use a helper function for truncation?

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/create.py [hunk 5]
```diff
@@ -95,9 +103,9 @@ def create_rules(cmd, cluster_subscription, cluster_resource_group_name, cluster
         default_rule_group_id,
         RULES_API
     )
-    put_rules(cmd, default_rule_group_id, default_rule_group_name, mac_region, azure_monitor_workspace_resource_id, cluster_name, default_rules_template, url, enable_windows_recording_rules, 2)
+    put_rules(cmd, default_rule_group_id, default_rule_group_name, mac_region, cluster_resource_id, azure_monitor_workspace_resource_id, cluster_name, default_rules_template, url, enable_windows_recording_rules, 2)
 
-    default_rule_group_name = "NodeAndKubernetesRecordingRulesRuleGroup-Win-{0}".format(cluster_name)
+    default_rule_group_name = "NodeAndKubernetesRecordingRulesRuleGroup-Win-{0}".format(cluster_name[:215])
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/create.py:108]: use a helper function for truncation?

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py [hunk 6]
```diff
@@ -23,27 +23,28 @@ def delete_rule(cmd, cluster_subscription, cluster_resource_group_name, default_
 
 
 def delete_rules(cmd, cluster_subscription, cluster_resource_group_name, cluster_name):
+    # limit rule group name to 260 characters
     delete_rule(
         cmd,
         cluster_subscription,
         cluster_resource_group_name,
-        "NodeRecordingRulesRuleGroup-{0}".format(cluster_name)
+        "NodeRecordingRulesRuleGroup-{0}".format(cluster_name[:232])
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py:31]: use a helper function for truncation?

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py [hunk 7]
```diff
@@ -23,27 +23,28 @@ def delete_rule(cmd, cluster_subscription, cluster_resource_group_name, default_
...
     )
     delete_rule(
         cmd,
         cluster_subscription,
         cluster_resource_group_name,
-        "KubernetesRecordingRulesRuleGroup-{0}".format(cluster_name)
+        "KubernetesRecordingRulesRuleGroup-{0}".format(cluster_name[:226])
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py:37]: use a helper function for truncation?

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py [hunk 8]
```diff
@@ -23,27 +23,28 @@ def delete_rule(cmd, cluster_subscription, cluster_resource_group_name, default_
...
     )
     delete_rule(
         cmd,
         cluster_subscription,
         cluster_resource_group_name,
-        "NodeRecordingRulesRuleGroup-Win-{0}".format(cluster_name)
+        "NodeRecordingRulesRuleGroup-Win-{0}".format(cluster_name[:228])
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py:43]: use a helper function for truncation?

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py [hunk 9]
```diff
@@ -23,27 +23,28 @@ def delete_rule(cmd, cluster_subscription, cluster_resource_group_name, default_
...
     )
     delete_rule(
         cmd,
         cluster_subscription,
         cluster_resource_group_name,
-        "NodeAndKubernetesRecordingRulesRuleGroup-Win-{0}".format(cluster_name)
+        "NodeAndKubernetesRecordingRulesRuleGroup-Win-{0}".format(cluster_name[:215])
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py:49]: use a helper function for truncation?

**Review by FumingZhang — COMMENTED** [review]
Queued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=82519989&view=results) to verify there's no regression introduced by the change. Test passed.

**Review by FumingZhang — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/create.py (hunk 2 above)
Comment (bragi92) [reply src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/create.py:58]: https://github.com/Azure/azure-cli/pull/27788/files

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/create.py (hunk 3 above)
Comment (bragi92) [reply src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/create.py:77]: https://github.com/Azure/azure-cli/pull/27788/files

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/create.py (hunk 5 above)
Comment (bragi92) [reply src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/create.py:108]: https://github.com/Azure/azure-cli/pull/27788/files

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py (hunk 6 above)
Comment (bragi92) [reply src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py:31]: https://github.com/Azure/azure-cli/pull/27788/files

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py (hunk 7 above)
Comment (bragi92) [reply src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py:37]: https://github.com/Azure/azure-cli/pull/27788/files

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py (hunk 8 above)
Comment (bragi92) [reply src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py:43]: https://github.com/Azure/azure-cli/pull/27788/files

File: src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py (hunk 9 above)
Comment (bragi92) [reply src/azure-cli/azure/cli/command_modules/acs/azuremonitormetrics/recordingrules/delete.py:49]: https://github.com/Azure/azure-cli/pull/27788/files

**Pull request merged**